File: content/snippets/scala3/language_basics_linked_lists.scala
object LinkedListBasics:
  case class Node(value: Int, next: Option[Node] = None)

  def printList(head: Option[Node]): Unit =
    val parts = Iterator.unfold(head) {
      case None       => None
      case Some(node) => Some((node.value.toString, node.next))
    }
    println(parts.mkString(" -> "))

  def prepend(head: Option[Node], value: Int): Node = Node(value, head)

  def removeValue(head: Option[Node], value: Int): Option[Node] = head match {
    case None                              => None
    case Some(node) if node.value == value => node.next
    case Some(node)                        => Some(node.copy(next = removeValue(node.next, value)))
  }

  @main def main(): Unit =
    var head: Option[Node] = None
    for (v <- List(3, 2, 1)) head = Some(prepend(head, v))  // 1 -> 2 -> 3
    printList(head)
    head = removeValue(head, 2)
    printList(head)  // 1 -> 3

File: content/snippets/go/language_basics_linked_lists.go
package main

import "fmt"

type Node struct {
	value int
	next  *Node
}

func printList(head *Node) {
	for cur := head; cur != nil; cur = cur.next {
		fmt.Print(cur.value)
		if cur.next != nil {
			fmt.Print(" -> ")
		}
	}
	fmt.Println()
}

func prepend(head *Node, value int) *Node {
	return &Node{value: value, next: head}
}

func removeValue(head *Node, value int) *Node {
	if head == nil {
		return nil
	}
	if head.value == value {
		return head.next
	}
	head.next = removeValue(head.next, value)
	return head
}

func main() {
	var head *Node
	for _, v := range []int{3, 2, 1} {
		head = prepend(head, v) // 1 -> 2 -> 3
	}
	printList(head)
	head = removeValue(head, 2)
	printList(head) // 1 -> 3
}
